File: src/restricted-proxy/find_collision.py
from web3 import Web3
import string
import random

def get_selector(func_sig):
    """Get first 4 bytes of keccak256 hash of the function signature"""
    return Web3.keccak(text=func_sig)[:4]

def random_string(length):
    """Generate random string of given length"""
    chars = string.ascii_letters + string.digits + '_'
    return ''.join(random.choice(chars) for _ in range(length))

target_selector = get_selector("changeWithdrawRate(uint8)")
print(f"Target selector: 0x{target_selector.hex()}")

# Try random strings until we find a collision
count = 0
while True:
    count += 1
    if count % 1000000 == 0:
        print(f"Tried {count} combinations...")
    
    # Generate random function name
    name = random_string(random.randint(1, 20))
    func_sig = f"{name}(uint256)"
    
    selector = get_selector(func_sig)
    if selector == target_selector:
        print(f"\nFound collision!")
        print(f"Original: changeWithdrawRate(uint8)")
        print(f"Colliding: {func_sig}")
        print(f"Selector: 0x{selector.hex()}")
        break

File: src/restricted-proxy/find_collision.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"sync/atomic"
	"time"

	"golang.org/x/crypto/sha3"
)

const (
	charset    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"
	numWorkers = 16
)

var (
	found      int32
	totalTries int64
)

func randomString(length int) string {
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

func getSelector(funcSig string) []byte {
	hash := sha3.NewLegacyKeccak256()
	hash.Write([]byte(funcSig))
	return hash.Sum(nil)[:4]
}

func worker(targetSelector []byte, resultChan chan string, wg *sync.WaitGroup) {
	defer wg.Done()

	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	for atomic.LoadInt32(&found) == 0 {
		// Generate random function name
		nameLen := r.Intn(20) + 1
		name := randomString(nameLen)
		funcSig := name + "(uint256)"

		atomic.AddInt64(&totalTries, 1)

		if tries := atomic.LoadInt64(&totalTries); tries%1000000 == 0 {
			fmt.Printf("Tried %d combinations...\n", tries)
		}

		selector := getSelector(funcSig)
		if string(selector) == string(targetSelector) {
			atomic.StoreInt32(&found, 1)
			resultChan <- funcSig
			return
		}
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())

	targetSelector := getSelector("changeWithdrawRate(uint8)")
	fmt.Printf("Target selector: 0x%x\n", targetSelector)

	resultChan := make(chan string)
	var wg sync.WaitGroup

	// Start workers
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go worker(targetSelector, resultChan, &wg)
	}

	// Wait for result
	result := <-resultChan
	fmt.Printf("\nFound collision!\n")
	fmt.Printf("Original: changeWithdrawRate(uint8)\n")
	fmt.Printf("Colliding: %s\n", result)
	fmt.Printf("Selector: 0x%x\n", targetSelector)
	fmt.Printf("Total tries: %d\n", atomic.LoadInt64(&totalTries))
}
